Framework/Kernel/src/Unit.cpp

```diff
@@ -644,56 +644,80 @@ void dSpacing::init() {
       }
     }
   }
+  // force the assumption that difc is positive in debug builds
+  assert(difc > 0.);
 }
 
 double dSpacing::singleToTOF(const double x) const {
   if (!isInitialized())
     throw std::runtime_error("dSpacingBase::singleToTOF called before object "
                              "has been initialized.");
-  return difa * x * x + difc * x + tzero;
+  if (difa == 0.)
+    return difc * x + tzero;
+  else
+    return difa * x * x + difc * x + tzero;
 }
 
+/**
+ * DIFA * d^2 + DIFC * d + T0 - TOF = 0
+ *
+ * Use the citardauq formula to solve quadratic in order to minimise loss of precision. citarqauq (quadratic spelled
+ * backwards) is an alternate formulation of the quadratic formula. DIFC and sqrt term are often similar and the
+ * "classic" quadratic formula involves calculating their difference in the numerator
+ *
+ *               2*(T0 - TOF)                                            (T0 - TOF)
+ * d = -------------------------------------------  =  ---------------------------------------------------
+ *     -DIFC -+ SQRT(DIFC^2 - 4*DIFA*(T0 - TOF))       0.5 * DIFC (-1 -+ SQRT(1 - 4*DIFA*(T0 - TOF)/DIFC^2)
+ *
+ * the variables in this formulation are the same as the quadratic formula
+ * a = difa      square term
+ * b = DIFC      linear term - assumed to be positive
+ * c = T0 - TOF  constant term
+ */
 double dSpacing::singleFromTOF(const double tof) const {
-  // DIFA * d^2 + DIFC * d + T0 - TOF = 0
-
-  // Use the citardauq formula to solve quadratic in order to minimise loss of precision. DIFC and sqrt term are often
-  // similar and the "classic" quadratic formula involves calculating their difference in the numerator
-
-  //               2*(T0 - TOF)                                            (T0 - TOF)
-  // d = -------------------------------------------  =  ---------------------------------------------------
-  //     -DIFC -+ SQRT(DIFC^2 - 4*DIFA*(T0 - TOF))       0.5 * DIFC (-1 -+ SQRT(1 - 4*DIFA*(T0 - TOF)/DIFC^2)
-
   // dealing with various edge cases
   if (!isInitialized())
     throw std::runtime_error("dSpacingBase::singleFromTOF called before object "
                              "has been initialized.");
   if (!toDSpacingError.empty())
     throw std::runtime_error(toDSpacingError);
-  double c = tzero - tof;
+
+  // force the assumption that difc is positive in debug builds
+  assert(difc > 0.);
+
   // handle special cases first...
-  // citardauq formula hides non-zero root if c=0
+
+  // don't need to solve a quadratic when difa==0
+  if (difa == 0.)
+    return (tof - tzero) / difc;
```

Comment:
I did think about whether to do this when I originally wrote the code. I decided not to in order to make the code cleaner and also because the performance hit of going through the full quadratic code seemed minimal. When difa=0, the sqrt is always sqrt(1) and this is quicker than sqrt on other values eg I timed sqrt(1) vs sqrt(M_PI) on MSVC and sqrt(1) was faster. Not sure if it's optimised or the sqrt algorithm just completes quicker on 1.

Even without this when I profiled doing unit conversions on large event workspaces as part of this change, this unit conversion code never showed up anywhere near the top

Reply:
Thank you for the results of the measurement. Have you checked adding in the flops for calculating the stuff that goes into the `std::sqrt` or just sqrt itself?